File: third-party/thrift/src/thrift/lib/cpp/transport/THttpClient.cpp
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <thrift/lib/cpp/transport/THttpClient.h>

#include <cstdlib>
#include <sstream>

#include <boost/algorithm/string.hpp>
#include <boost/lexical_cast.hpp>

#include <thrift/lib/cpp/transport/TSocket.h>

namespace apache {
namespace thrift {
namespace transport {

using std::string;

const string THttpClient::kAcceptHeader = "Accept";
const string THttpClient::kConnectionHeader = "Connection";
const string THttpClient::kContentLengthHeader = "Content-Length";
const string THttpClient::kContentTypeHeader = "Content-Type";
const string THttpClient::kHostHeader = "Host";
const string THttpClient::kTransferEncodingHeader = "Transfer-Encoding";
const string THttpClient::kUserAgentHeader = "User-Agent";

THttpClient::THttpClient(
    const std::shared_ptr<TTransport>& transport,
    const string& host,
    const string& path)
    : THttpTransport(transport),
      host_(host),
      path_(path),
      connectionClosedByServer_(false) {
  setHeader(kUserAgentHeader, "C++/THttpClient");
  setHeader(kContentTypeHeader, "application/x-thrift");
  setHeader(kAcceptHeader, "application/x-thrift");
}

THttpClient::THttpClient(const string& host, int port, const string& path)
    : THttpTransport(std::shared_ptr<TTransport>(new TSocket(host, port))),
      host_(host),
      path_(path),
      connectionClosedByServer_(false) {
  setHeader(kUserAgentHeader, "C++/THttpClient");
  setHeader(kContentTypeHeader, "application/x-thrift");
  setHeader(kAcceptHeader, "application/x-thrift");
}

THttpClient::~THttpClient() {}

void THttpClient::beginParsingHeaders() {
  responseHeaders_.clear();
}

void THttpClient::parseHeader(char* header) {
  const char* colon = strchr(header, ':');
  if (colon == nullptr || colon == header) {
    return;
  }
  const char* value = colon + 1;
  while (*value == ' ') {
    value++;
  }

  string name(header, colon - header);

  // If the same header appears twice, we simply overwrite it
  responseHeaders_[name] = value;

  if (boost::iequals(name, kTransferEncodingHeader)) {
    if (boost::iequals(value, "chunked")) {
      chunked_ = true;
    }
  } else if (boost::iequals(name, kContentLengthHeader)) {
    chunked_ = false;
    contentLength_ = atoi(value);
  } else if (boost::iequals(name, kConnectionHeader)) {
    if (boost::iequals(value, "close")) {
      connectionClosedByServer_ = true;
    }
  }
}

bool THttpClient::parseStatusLine(char* status) {
  const char* http = status;

  // Skip over the "HTTP/<version>" string.
  // TODO: we should probably check that the version is 1.0 or 1.1
  const char* code = strchr(http, ' ');
  if (code == nullptr) {
    throw TTransportException(string("Bad Status: ") + status);
  }

  // RFC 2616 requires exactly 1 space between the HTTP version and the status
  // code. Skip over it.
  ++code;

  // RFC 2616 requires exactly 1 space between the the status code and reason
  // phrase.
  const char* endCode = strchr(code, ' ');
  if (endCode == nullptr) {
    throw TTransportException(string("Bad Status: ") + status);
  }

  try {
    statusCode_ = boost::lexical_cast<uint16_t>(string(code, endCode - code));
    if (statusCode_ == 100) {
      // HTTP 100 = continue, just keep reading
      return false;
    }
    return true;
  } catch (boost::bad_lexical_cast&) {
    throw TTransportException(string("Bad Status: ") + status);
  }
}

void THttpClient::endParsingHeaders() {
  if (statusCode_ != 200) {
    std::stringstream msg;
    msg << "Server returned HTTP status code " << statusCode_;
    throw TTransportException(msg.str());
  }
}

void THttpClient::flush() {
  if (connectionClosedByServer_) {
    close();
    open();
    connectionClosedByServer_ = false;
  }

  // Fetch the contents of the write buffer
  uint8_t* buf;
  uint32_t len;
  writeBuffer_.getBuffer(&buf, &len);

  // Construct the HTTP header
  std::ostringstream h;
  h << "POST " << path_ << " HTTP/1.1" << CRLF << kHostHeader << ": " << host_
    << CRLF;
  for (const auto& header : requestHeaders_) {
    h << header.first << ": " << header.second << CRLF;
  }
  h << kContentLengthHeader << ": " << len << CRLF << CRLF;
  string header = h.str();

  // Write the header, then the data, then flush
  transport_->write((const uint8_t*)header.c_str(), header.size());
  transport_->write(buf, len);
  transport_->flush();

  // Reset the buffer
  writeBuffer_.resetBuffer();
}

void THttpClient::setHeader(const std::string& name, const std::string& value) {
  requestHeaders_[name] = value;
}

void THttpClient::setUserAgent(const std::string& userAgent) {
  setHeader(kUserAgentHeader, userAgent);
}

} // namespace transport
} // namespace thrift
} // namespace apache

File: third-party/thrift/src/thrift/lib/py/transport/THttpClient.py
# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# pyre-unsafe

from __future__ import absolute_import, division, print_function, unicode_literals

import http.client as http_client
import os
import socket
import sys
import warnings
from io import BytesIO as StringIO
from urllib import parse

from thrift.transport.TTransport import TTransportBase, TTransportException


class THttpClient(TTransportBase):
    """Http implementation of TTransport base."""

    def __init__(self, uri_or_host, port=None, path=None, ssl_context=None):
        """THttpClient supports two different types constructor parameters.

        THttpClient(host, port, path) - deprecated
        THttpClient(uri)

        Only the second supports https."""

        if port is not None:
            warnings.warn(
                "Please use the THttpClient('http://host:port/path') syntax",
                DeprecationWarning,
                stacklevel=2,
            )
            self.host = uri_or_host
            self.http_host = self.host
            self.port = port
            assert path
            self.path = path
            self.scheme = "http"
        else:
            parsed = parse.urlparse(uri_or_host)
            self.scheme = parsed.scheme
            assert self.scheme in ("http", "https")
            if self.scheme == "http":
                self.port = parsed.port or http_client.HTTP_PORT
            elif self.scheme == "https":
                self.port = parsed.port or http_client.HTTPS_PORT
            self.host = parsed.hostname
            self.http_host = parsed.netloc
            self.path = parsed.path
            if parsed.query:
                self.path += "?%s" % parsed.query
        self.__wbuf = StringIO()
        self.__http = None
        self.__timeout = None
        self.__custom_headers = None
        self.ssl_context = ssl_context

    def open(self):
        if self.scheme == "http":
            self.__http = http_client.HTTPConnection(
                self.host, self.port, timeout=self.__timeout
            )
        else:
            self.__http = http_client.HTTPSConnection(
                self.host, self.port, context=self.ssl_context, timeout=self.__timeout
            )

    def close(self):
        self.__http.close()
        self.__http = None

    def isOpen(self):
        return self.__http is not None

    def setTimeout(self, ms):
        if ms is None:
            self.__timeout = None
        else:
            self.__timeout = ms / 1000.0

    def setCustomHeaders(self, headers):
        self.__custom_headers = headers

    def setCustomHeader(self, name, value):
        if self.__custom_headers is None:
            self.__custom_headers = {}
        self.__custom_headers[name] = value

    def read(self, sz):
        return self.response.read(sz)

    def write(self, buf):
        self.__wbuf.write(buf)

    def flush(self):
        if self.isOpen():
            self.close()
        self.open()

        # Pull data out of buffer
        data = self.__wbuf.getvalue()
        self.__wbuf = StringIO()

        # HTTP request
        self.__http.putrequest("POST", self.path, skip_host=True)

        if not self.__custom_headers or "Host" not in self.__custom_headers:
            self.__http.putheader("Host", self.http_host)

        self.__http.putheader("Content-Type", "application/x-thrift")
        self.__http.putheader("Content-Length", str(len(data)))

        if not self.__custom_headers or "User-Agent" not in self.__custom_headers:
            user_agent = "Python/THttpClient"
            script = os.path.basename(sys.argv[0])
            if script:
                user_agent = "%s (%s)" % (user_agent, parse.quote(script))
            self.__http.putheader("User-Agent", user_agent)

        if self.__custom_headers:
            if sys.version_info[0] >= 3:
                custom_headers_iter = self.__custom_headers.items()
            else:
                custom_headers_iter = self.__custom_headers.items()
            for key, val in custom_headers_iter:
                self.__http.putheader(key, val)

        try:
            self.__http.endheaders()

            # Write payload
            self.__http.send(data)
        except socket.gaierror as e:
            raise TTransportException(TTransportException.NOT_OPEN, str(e))
        except Exception as e:
            raise TTransportException(TTransportException.UNKNOWN, str(e))

        # Get reply to flush the request
        self.response = self.__http.getresponse()
        self.code = self.response.status
        self.headers = self.response.getheaders()
